encoding/codecv0.go

```
@@ -0,0 +1,335 @@
+		// build data hash
+		daChunk, err := o.NewDAChunk(chunk, totalL1MessagePoppedBeforeChunk)
+		if err != nil {
+			return nil, err
+		}
+		totalL1MessagePoppedBeforeChunk += chunk.NumL1Messages(totalL1MessagePoppedBeforeChunk)
+		daChunkHash, err := daChunk.Hash()
+		if err != nil {
+			return nil, err
+		}
+		dataBytes = append(dataBytes, daChunkHash.Bytes()...)
+	}
+
+	// compute data hash
+	dataHash := crypto.Keccak256Hash(dataBytes)
+
+	// skipped L1 messages bitmap
+	bitmapBytes, totalL1MessagePoppedAfter, err := constructSkippedBitmap(batch.Index, batch.Chunks, batch.TotalL1MessagePoppedBefore)
+	if err != nil {
+		return nil, err
+	}
+
+	daBatch := DABatchV0{
+		DABatchBase: DABatchBase{
+			Version:                uint8(CodecV0),
+			BatchIndex:             batch.Index,
+			L1MessagePopped:        totalL1MessagePoppedAfter - batch.TotalL1MessagePoppedBefore,
+			TotalL1MessagePopped:   totalL1MessagePoppedAfter,
+			DataHash:               dataHash,
+			ParentBatchHash:        batch.ParentBatchHash,
+			SkippedL1MessageBitmap: bitmapBytes,
+		},
+	}
+
+	return &daBatch, nil
+}
+
+// NewDABatchWithExpectedBlobVersionedHashes creates a DABatch from the provided Batch.
+// It also checks if the blob versioned hashes are as expected.
+func (o *DACodecV0) NewDABatchWithExpectedBlobVersionedHashes(batch *Batch, hashes []common.Hash) (DABatch, error) {
+	daBatch, err := o.NewDABatch(batch)
+	if err != nil {
+		return nil, err
+	}
+
+	if reflect.DeepEqual(daBatch.BlobVersionedHashes(), hashes) {
+		return nil, errors.New("blob versioned hashes do not match")
+	}
+
+	return daBatch, nil
+}
+
+// NewDABatchFromBytes decodes the given byte slice into a DABatch.
+func (o *DACodecV0) NewDABatchFromBytes(data []byte) (DABatch, error) {
+	if len(data) < 89 {
+		return nil, fmt.Errorf("insufficient data for DABatch, expected at least 89 bytes but got %d", len(data))
+	}
+
+	b := &DABatchV0{
+		DABatchBase: DABatchBase{
+			Version:                data[0],
```

we should return an error here if it doesn't match the codec's version